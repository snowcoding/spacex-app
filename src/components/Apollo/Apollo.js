import React from 'react'
import ApolloClient from 'apollo-boost'
import { ApolloProvider } from 'react-apollo'

const client = new ApolloClient({
  uri: 'https://api.spacex.land/graphql/',
})

export default function Apollo(props) {
  return <ApolloProvider client={client}>{props.children}</ApolloProvider>
}
