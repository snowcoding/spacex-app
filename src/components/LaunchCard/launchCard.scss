@import '../../globals.scss';

i.fas.fa-times-circle.failure {
  color: red;
}

i.fas.fa-check-circle.pass {
  color: green;
}

i.far.fa-space-shuttle,i.fal.fa-satellite,i.fal.fa-dragon {
  color: #555;
}

div.flag-group {
  display: flex;
  justify-content: 'space-around';
}

div.launch-date, div.rocket-row {
  font-size: 1.3em;
  align-self: flex-start;
}

div.payload {
  font-size: 2em;
}

span.date {
  margin-left: 10px;
}

span.rocket-text {
  margin-left: 7px;
}

h2.mission-name {
  color: $mui-primary;
}
