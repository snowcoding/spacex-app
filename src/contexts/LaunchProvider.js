import React, { createContext, useReducer } from 'react'

export const LaunchContext = createContext()

const reducer = (state, action) => {
  console.log('Launch Filter State: ', state)
  console.log('Launch Filter Action: ', action)
  switch (action.type) {
    case 'changeFilterDateRange':
      return {
        ...state,
        filterDateRange: [...action.filterDateRange],
      }

    default:
      throw new Error()
  }
}

export const changeFilterDateRange = filterDateRange => {
  return {
    type: 'changeFilterDateRange',
    filterDateRange,
  }
}

const LaunchProvider = props => {
  const initialState = {
    filterDateRange: [2006, 2020],
  }

  const [state, dispatch] = useReducer(reducer, initialState)

  return (
    <LaunchContext.Provider value={[state, dispatch]}>
      {props.children}
    </LaunchContext.Provider>
  )
}

export default LaunchProvider
