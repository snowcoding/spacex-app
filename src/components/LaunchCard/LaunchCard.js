import React, { useContext } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import Flag from 'react-world-flags'
import './launchCard.scss'
import {
  LaunchContext,
  changeLaunchCardDetails,
} from '../../contexts/LaunchProvider'

const useStyles = makeStyles({
  card: {
    minWidth: 180,
    maxWidth: 300,
    height: 250,
    border: '2px solid rgba(0,0,0,0)',
    transition: 'all 0.2s',
    '&:hover': {
      boxShadow: '0px 2px 7px #888',
    },
    backgroundColor: '#FFF',
  },
  cardContent: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
    alignItems: 'center',
    height: 250,
  },
  flag: {
    height: 27,
    marginLeft: 5,
    boxShadow: '1px 0px 4px #888',
  },
})

export default function LaunchCard(props) {
  const classes = useStyles()
  const [state, dispatch] = useContext(LaunchContext)

  const { rocketName, payloads, missionName, date, success, launchId } = props

  const uniqueNations = payloads
    ? [...new Set(payloads.map(pl => pl.nationality))]
    : []

  const flagCodeLUT = {
    'United States': 'USA',
    Argentina: 'ARG',
    Bangladesh: 'BGD',
    Bulgaria: 'BGR',
    Canada: 'CAN',
    France: 'FRA',
    'Hong Kong': 'HKG',
    Indonesia: 'IDN',
    Israel: 'ISR',
    Japan: 'JPN',
    Luxembourg: 'LUX',
    Qatar: 'QAT',
    'South Korea': 'KOR',
    Spain: 'ESP',
    Thailand: 'THA',
    Turkmenistan: 'TKM',
    Taiwan: 'TWN',
    'Saudi Arabia': 'SAU',
    'United Kingdom': 'GBR',
    Malaysia: 'MYS',
  }

  const openLaunchCardDetails = () => {
    dispatch(changeLaunchCardDetails(launchId))
  }

  return (
    <Card className={classes.card} onClick={openLaunchCardDetails}>
      <CardContent className={classes.cardContent}>
        <div className='flag-group'>
          {uniqueNations.length !== 0 ? (
            uniqueNations.map((nation, ind) =>
              nation && flagCodeLUT[nation] ? (
                <Flag
                  key={`${ind}+${launchId}`}
                  className={classes.flag}
                  code={flagCodeLUT[nation]}
                />
              ) : (
                <Flag key={`${ind}+${launchId}`} className={classes.flag} code={'USA'} />
              )
            )
          ) : (
            <Flag className={classes.flag} code={'USA'} />
          )}
        </div>

        <h2 className='mission-name'>
          {missionName ? missionName.substring(0, 8) : null}
        </h2>

        <div className='launch-date'>
          {!success ? (
            <i className='fas fa-times-circle failure' />
          ) : (
            <i className='fas fa-check-circle pass' />
          )}
          <span className='date'>{date ? date.split('T')[0] : null}</span>
        </div>

        <div className='rocket-row'>
          <i className='far fa-space-shuttle' />
          <span className='rocket-text'>{rocketName}</span>
        </div>

        <div className='payload'>
          {/* Map through the payloads array, if it has more than 4 items, 
          render all satellites, otherwise, render the appropriate icon */}
          {payloads && payloads.length > 3
            ? [0, 1, 2, 3].map((cv, ind) => (
                <i key={`${ind}+${launchId}`} className='fal fa-satellite' />
              ))
            : payloads.map((pl, ind) =>
                pl.payload_type === 'Satellite' ? (
                  <i key={`${ind}+${launchId}`} className='fal fa-satellite' />
                ) : (
                  <i key={`${ind}+${launchId}`} className='fal fa-dragon' />
                )
              )}
        </div>
      </CardContent>
    </Card>
  )
}
