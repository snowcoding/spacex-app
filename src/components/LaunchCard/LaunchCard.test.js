import React from 'react'
import ReactDOM from 'react-dom'
import LaunchCard from './LaunchCard'
import Apollo from '../Apollo/Apollo'
import LaunchProvider from '../../contexts/LaunchProvider'
import renderer from 'react-test-renderer';

it('snapshot test of launch card', () => {
  const rocketName = 'Falcon'
  const payloads = [{ payload_type: 'Satellite', nationality: 'United States' }]
  const missionName = 'RockLauncher'
  const date = '2019-07-11'
  const success = true
  const launchId = 77
  
  const tree = renderer
    .create(      <LaunchProvider>
      <LaunchCard
        rocketName={rocketName}
        payloads={payloads}
        missionName={missionName}
        date={date}
        success={success}
        launchId={launchId}
      />
    </LaunchProvider>
)
    .toJSON();
  expect(tree).toMatchSnapshot();
});


it('renders without crashing', () => {
  const div = document.createElement('div')
  const rocketName = 'Falcon'
  const payloads = [{ payload_type: 'Satellite', nationality: 'United States' }]
  const missionName = 'RockLauncher'
  const date = '2019-07-11'
  const success = true
  const launchId = 77

  ReactDOM.render(
    <Apollo>
      <LaunchProvider>
        <LaunchCard
          rocketName={rocketName}
          payloads={payloads}
          missionName={missionName}
          date={date}
          success={success}
          launchId={launchId}
        />
      </LaunchProvider>
    </Apollo>,
    div
  )
  ReactDOM.unmountComponentAtNode(div)
})
