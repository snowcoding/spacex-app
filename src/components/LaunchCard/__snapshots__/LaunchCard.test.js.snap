// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test of launch card 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"
  onClick={[Function]}
>
  <div
    className="MuiCardContent-root makeStyles-cardContent-2"
  >
    <div
      className="flag-group"
    >
      <img
        className="makeStyles-flag-3"
        src="data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' viewBox='0 0 7410 3900'%3E %3Crect width='7410' height='3900' fill='%23b22234'/%3E %3Cpath d='M0,450H7410m0,600H0m0,600H7410m0,600H0m0,600H7410m0,600H0' stroke='%23fff' stroke-width='300'/%3E %3Crect width='2964' height='2100' fill='%233c3b6e'/%3E %3Cg fill='%23fff'%3E %3Cg id='s18'%3E %3Cg id='s9'%3E %3Cg id='s5'%3E %3Cg id='s4'%3E %3Cpath id='s' d='M247,90 317.534230,307.082039 132.873218,172.917961H361.126782L176.465770,307.082039z'/%3E %3Cuse xlink:href='%23s' y='420'/%3E %3Cuse xlink:href='%23s' y='840'/%3E %3Cuse xlink:href='%23s' y='1260'/%3E %3C/g%3E %3Cuse xlink:href='%23s' y='1680'/%3E %3C/g%3E %3Cuse xlink:href='%23s4' x='247' y='210'/%3E %3C/g%3E %3Cuse xlink:href='%23s9' x='494'/%3E %3C/g%3E %3Cuse xlink:href='%23s18' x='988'/%3E %3Cuse xlink:href='%23s9' x='1976'/%3E %3Cuse xlink:href='%23s5' x='2470'/%3E %3C/g%3E %3C/svg%3E"
      />
    </div>
    <h2
      className="mission-name"
    >
      RockLaun
    </h2>
    <div
      className="launch-date"
    >
      <i
        className="fas fa-check-circle pass"
      />
      <span
        className="date"
      >
        2019-07-11
      </span>
    </div>
    <div
      className="rocket-row"
    >
      <i
        className="far fa-space-shuttle"
      />
      <span
        className="rocket-text"
      >
        Falcon
      </span>
    </div>
    <div
      className="payload"
    >
      <i
        className="fal fa-satellite"
      />
    </div>
  </div>
</div>
`;
